example/timer/odp_timer_accuracy.c

```diff
@@ -1107,63 +1177,92 @@ int main(int argc, char *argv[])
 	/* Configure scheduler */
 	odp_schedule_config(NULL);
 
-	test_global.timer_ctx = calloc(test_global.alloc_timers,
-				       sizeof(timer_ctx_t));
+	odp_shm_t shm = ODP_SHM_INVALID;
+	void *addr;
+	uint64_t size = sizeof(test_shm_t);
 
-	if (test_global.timer_ctx == NULL) {
-		printf("Timer context table calloc failed.\n");
-		ret = -1;
-		goto quit;
-	}
+	size += test_global.alloc_timers * sizeof(timer_ctx_t);
 
 	if (test_global.opt.output) {
-		odp_shm_t shm;
-		void *addr;
-		uint64_t size = test_global.tot_timers * sizeof(test_log_t);
-
 		test_global.file = fopen(test_global.filename, "w");
 		if (test_global.file == NULL) {
 			printf("Failed to open output file %s: %s\n",
 			       test_global.filename, strerror(errno));
 			ret = -1;
 			goto quit;
 		}
+		size += (test_global.tot_timers - test_global.warmup_timers) * sizeof(test_log_t);
+	}
 
-		shm = odp_shm_reserve("timer_accuracy_log", size,
-				      sizeof(test_log_t), 0);
+	shm = odp_shm_reserve("timer_accuracy", size,
+			      ODP_CACHE_LINE_SIZE, ODP_SHM_SINGLE_VA);
 
-		if (shm == ODP_SHM_INVALID) {
-			printf("Test log alloc failed.\n");
-			ret = -1;
-			goto quit;
-		}
-
-		addr = odp_shm_addr(shm);
-		memset(addr, 0, size);
-		test_global.log = addr;
-		test_global.log_shm = shm;
+	if (shm == ODP_SHM_INVALID) {
+		printf("Test log alloc failed.\n");
+		ret = -1;
+		goto quit;
 	}
 
+	addr = odp_shm_addr(shm);
+	memset(addr, 0, size);
+	test_global.shm = addr;
+	test_global.shm->timer_ctx = (timer_ctx_t *)((test_shm_t *)addr + 1);
+	if (test_global.opt.output)
+		test_global.shm->log =
+			(test_log_t *)&test_global.shm->timer_ctx[test_global.alloc_timers];
+
+	odph_thread_t thread_tbl[MAX_WORKERS];
+	int num_workers;
+	odp_cpumask_t cpumask;
+	char cpumaskstr[ODP_CPUMASK_STR_SIZE];
+	odph_thread_common_param_t thr_common;
+	odph_thread_param_t thr_param;
+
+	memset(thread_tbl, 0, sizeof(thread_tbl));
+
+	num_workers = MAX_WORKERS;
+	if (test_global.opt.cpu_count && test_global.opt.cpu_count < MAX_WORKERS)
+		num_workers = test_global.opt.cpu_count;
+	num_workers = odp_cpumask_default_worker(&cpumask, num_workers);
+	odp_cpumask_to_str(&cpumask, cpumaskstr, sizeof(cpumaskstr));
+
+	printf("num worker threads: %i\n", num_workers);
+	printf("first CPU:          %i\n", odp_cpumask_first(&cpumask));
+	printf("cpu mask:           %s\n", cpumaskstr);
+
+	odp_barrier_init(&test_global.shm->barrier, num_workers + 1);
+	odp_atomic_init_u64(&test_global.shm->events, 0);
+
+	odph_thread_param_init(&thr_param);
+	thr_param.start = run_test;
+	thr_param.arg = (void *)&test_global;
+	thr_param.thr_type = ODP_THREAD_WORKER;
+
+	odph_thread_common_param_init(&thr_common);
+	thr_common.instance = instance;
+	thr_common.cpumask = &cpumask;
+	thr_common.share_param = 1;
+
+	odph_thread_create(thread_tbl, &thr_common, &thr_param, num_workers);
+
 	ret = start_timers(&test_global);
 	if (ret)
 		goto quit;
 
-	run_test(&test_global);
-
+	odp_barrier_wait(&test_global.shm->barrier);
```

Comment:
With multiple threads (and with the barrier synchronizing the start up), it would be safer to split start_timers() into two parts:
1) create all timers (exit if returns fail)
2) start all workers, wait until those are running (barrier)
3) start timers (get current time + odp_timer_start() loop)

Main point is that barrier after timer start prevents all workers from handling the first timeouts, if one thread is slower to start / got interrupted before the barrier.